src/libstd/ffi/c_str.rs

```diff
@@ -255,6 +255,34 @@ impl CString {
         Box::into_raw(self.inner) as *mut libc::c_char
     }
 
+    /// Converts the `CString` into a `String` if it contains valid Unicode data.
+    ///
+    /// On failure, ownership of the original `CString` is returned.
+    #[unstable(feature = "convert", reason = "recently added", issue = "27704")]
+    pub fn into_string(self) -> Result<String, CString> {
```

Comment:
Unlike `OsString`'s `into_string` method this actually has some extra contextual information to give back (e.g. the `str::Utf8Error`). Along those lines this may want to return its own custom error type (e.g. the same as `FromUtf8Error` but with a `CString` instead of `String`.